controllers/machinehealthcheck_targets.go

```diff
@@ -0,0 +1,259 @@
+		return nil, nil
+	}
+
+	targets := []healthCheckTarget{}
+	for k := range machines {
+		target := healthCheckTarget{
+			MHC:     mhc,
+			Machine: &machines[k],
+		}
+		node, err := r.getNodeFromMachine(clusterClient, target.Machine)
+		if err != nil {
+			if !apierrors.IsNotFound(err) {
+				return nil, errors.Wrap(err, "error getting node")
+			}
+			// a node with only a name represents a
+			// not found node in the target
+			node.Name = machines[k].Status.NodeRef.Name
+		}
+		target.Node = node
+		targets = append(targets, target)
+	}
+	return targets, nil
+}
+
+//getMachinesFromMHC fetches Machines matched by the MachineHealthCheck's
+// label selector
+func (r *MachineHealthCheckReconciler) getMachinesFromMHC(mhc *clusterv1.MachineHealthCheck) ([]clusterv1.Machine, error) {
+	selector, err := metav1.LabelSelectorAsSelector(&mhc.Spec.Selector)
+	if err != nil {
+		return nil, errors.New("failed to build selector")
+	}
+
+	options := client.ListOptions{
+		LabelSelector: selector,
+		Namespace:     mhc.GetNamespace(),
+	}
+	machineList := &clusterv1.MachineList{}
+	if err := r.Client.List(context.Background(), machineList, &options); err != nil {
+		return nil, errors.Wrap(err, "failed to list machines")
+	}
+	return machineList.Items, nil
+}
+
+// getNodeFromMachine fetches the node from a local or remote cluster for a
+// given machine.
+func (r *MachineHealthCheckReconciler) getNodeFromMachine(clusterClient client.Client, machine *clusterv1.Machine) (*corev1.Node, error) {
+	if machine.Status.NodeRef == nil {
+		return nil, nil
+	}
+
+	node := &corev1.Node{}
+	nodeKey := types.NamespacedName{
+		Namespace: machine.Status.NodeRef.Namespace,
+		Name:      machine.Status.NodeRef.Name,
+	}
+	err := clusterClient.Get(context.TODO(), nodeKey, node)
+	return node, err
+}
+
+// healthCheckTargets health checks a slice of targets
+// and gives a data to measure the average health
+func (r *MachineHealthCheckReconciler) healthCheckTargets(targets []healthCheckTarget, logger logr.Logger, timeoutForMachineToHaveNode time.Duration) (int, []healthCheckTarget, []time.Duration) {
+	var nextCheckTimes []time.Duration
+	var needRemediationTargets []healthCheckTarget
+	var currentHealthy int
+
+	for _, t := range targets {
+		logger = logger.WithValues("Target", t.string())
+		logger.V(3).Info("Health checking target")
+		needsRemediation, nextCheck := t.needsRemediation(logger, timeoutForMachineToHaveNode)
+
+		if needsRemediation {
+			needRemediationTargets = append(needRemediationTargets, t)
+			continue
+		}
+
+		if nextCheck > 0 {
+			logger.V(3).Info("Target is likely to go unhealthy", "timeUntilUnhealthy", nextCheck.Truncate(time.Second).String())
+			r.recorder.Eventf(
+				t.Machine,
+				corev1.EventTypeNormal,
+				EventDetectedUnhealthy,
+				"Machine %v has unhealthy node %v",
+				t.string(),
+				t.nodeName(),
+			)
+			nextCheckTimes = append(nextCheckTimes, nextCheck)
+			continue
+		}
+
+		if t.Machine.DeletionTimestamp.IsZero() {
```

Comment:
Do we need to exclude machines that are being deleted from the entirety of MHC?

Reply:
I don't think having them in breaks the logic at all, but it could be an optimisation to remove them before health checking (or short circuit in `needsRemediation`) since there's no point checking the conditions on a node we know is going away already. We need to count them in the targets so that the logic for calculating the `maxUnhealthy` will work